napari/utils/io.py

```diff
@@ -48,3 +50,168 @@ def imsave(filename: str, data: np.ndarray):
+        not impose any format requirements on the csv, and data will always be
+        returned.  If ``any``, csv must be recognized as one of the valid layer
+        data formats, otherwise a ``ValueError`` will be raised.  If a specific
+        layer type string, then a ``ValueError`` will be raised if the column
+        names are not of the predicted format.
+
+    Returns
+    -------
+    (data, column_names, layer_type) : Tuple[np.array, List[str], str]
+        The table data and column names from the CSV file, along with the
+        detected layer type (string).
+
+    Raises
+    ------
+    ValueError
+        If the column names do not match the format requested by
+        ``require_type``.
+    """
+    _warn_about_deprecated_import('read_csv')
+    import napari_builtins.io
+
+    return napari_builtins.io.read_csv(filename, require_type)
+
+
+def csv_to_layer_data(
+    path: str, require_type: Optional[str] = None
+) -> Optional[FullLayerData]:
+    """Return layer data from a CSV file if detected as a valid type.
+
+    Parameters
+    ----------
+    path : str
+        Path of file to open
+    require_type : str, optional
+        The desired layer type. If provided, should be one of the keys in
+        ``csv_reader_functions`` or the string "any".  If ``None``,
+        unrecognized CSV files will simply return ``None``.  If ``any``,
+        unrecognized CSV files will raise a ``ValueError``.  If a specific
+        layer type string, then a ``ValueError`` will be raised if the column
+        names are not of the predicted format.
+
+    Returns
+    -------
+    layer_data : tuple, or None
+        3-tuple ``(array, dict, str)`` (points data, metadata, layer_type) if
+        CSV is recognized as a valid type.
+
+    Raises
+    ------
+    ValueError
+        If ``require_type`` is not ``None``, but the CSV is not detected as a
+        valid data format.
+    """
+    _warn_about_deprecated_import('csv_to_layer_data')
+    import napari_builtins.io
+
+    return napari_builtins.io.csv_to_layer_data(path, require_type)
+
+
+def read_zarr_dataset(path):
+    """Read a zarr dataset, including an array or a group of arrays.
+
+    Parameters
+    ----------
+    path : str
+        Path to directory ending in '.zarr'. Path can contain either an array
+        or a group of arrays in the case of multiscale data.
+
+    Returns
+    -------
+    image : array-like
+        Array or list of arrays
+    shape : tuple
+        Shape of array or first array in list
+    """
+    _warn_about_deprecated_import('read_zarr_dataset')
+    import napari_builtins.io
+
+    return napari_builtins.io.read_zarr_dataset(path)
+
+
+def _warn_about_deprecated_import(name):
+    warnings.warn(
+        trans._(
+            '{name} was moved from napari.utils.io in 0.4.17. Import it from napari_builtins.io instead.',
+            deferred=True,
+            name=name,
+        ),
+        DeprecationWarning,
+        stacklevel=3,
+    )
```

Comment:
why use this approach instead of something like:

```suggestion


def __getattr__(name):

    if name in {'imsave_extensions', 'write_csv', 'read_csv', 'csv_to_layer_data', 'read_zarr_dataset'}:
        warnings.warn(
            trans._(
                '{name} was moved from napari.utils.io in 0.4.17. Import it from napari_builtins.io instead.',
                deferred=True,
                name=name,
            ),
            DeprecationWarning,
            stacklevel=3,
        )        
        import napari_builtins.io

        return getattr(napari_builtins.io, "name")

    raise AttributeError(f"module {__name__} has no attribute {name}")

```

Advantages of my proposition:
1) Warnings are on access, not on call
2) no need for sync documentation as a user will work on original functions.

The only possible disadvantage is when the user depends on the function `__module__` attribute.

Reply:
Does the stacklevel=... on import works now ? As far as I remember the number of stacks were incorrect, and it was not showing useful.

Maybe it's fixed now, IDK. 

Anyway the stacklevel should be 2 if this get added, or it will show one frame too high.

To be sure I would also show on import and call , via decorating what is returned on `return getattr(napari_builtins.io, "name")`, imports warnings may also get gobbled up at napari startup.

Reply:
I have fixed stacklevel. And napari uses `napari_builtins`

Reply:
Didn't think about doing this, but I love it. If you can just fix your suggestion to add a `str` type annotation to `name` and use name directly instead of `"name"`, then I'll just commit the suggestion as is. Alternatively, create a PR from this or just directly on top of `main`.

Ideally, we should only warn on import, but as @Carreau mentioned I'm a bit worried about napari's (troubling) behavior of silencing deprecation warnings. When using these changes (or with @Czaki's suggestions) with an npe2  plugin that uses one of these deprecated functions, I didn't actually see any warnings anywhere, so I'm still concerned by that, but maybe that's a separate issue.

Reply:
Fixed bug in code.

You does not see any warning because python silence DeprecationWarnings by defaults. This is reason why we mostly uses FutureWarnings.

Reply:
Ah, I forgot Python's distinction. Based on [the official docs](https://docs.python.org/3/library/exceptions.html#DeprecationWarning), it seems like `DeprecationWarning` is the right choice here in terms of intent - this is a warning for people developing against napari's API not for end users.